### PR TITLE
Feat/filters

### PR DESCRIPTION
**Pull request by alepefe-vizz**
- test(api): Clear mocks after each test in app.spec.ts
- feat(api): Add SearchDataWidgetParams in order to allow users to filter the data returned allong with the sections' widgets
- feat: Add initial Page Filters implementation (Contract, controller, service, tests)

### CODE REVIEW

File: api/data/filters/filters.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      "Romania",
+      "Slovakia",
+      "Slovenia",
+      "Spain",
+      "Sweden",
+      "The Netherlands"
+    ],
+    "label": {
+      "selected": "Country is",
+      "unSelected": "All Countries"
```
Comment (alexeh) [inline api/data/filters/filters.json:36]: Maybe we are missing one extra label for when multiple values are selected?

Smth like: `Countries are...`

But not sure

File: shared/schemas/search-widget-data-params.schema.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+export const SearchWidgetDataParamsSchema = z.object({
+  filters: z
+    .array(
+      z.object({
+        name: z.string(),
+        operator: z.enum(
+          VALID_SEARCH_WIDGET_DATA_OPERATORS as [
+            WidgetDataFilterOperatorType,
+            ...WidgetDataFilterOperatorType[],
```
Comment (alexeh) [inline shared/schemas/search-widget-data-params.schema.ts:15]: Can we simplify this? It's over-complicated

File: shared/dto/global/search-widget-data-params.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  EQUALS: '=',
+  NOT_EQUALS: '!=',
+} as const;
+
+export const VALID_SEARCH_WIDGET_DATA_OPERATORS = Object.values(
+  SEARCH_WIDGET_DATA_OPERATORS,
+);
+
+export type WidgetDataFilterOperatorType =
+  (typeof SEARCH_WIDGET_DATA_OPERATORS)[keyof typeof SEARCH_WIDGET_DATA_OPERATORS];
```
Comment (alexeh) [inline shared/dto/global/search-widget-data-params.ts:11]: I'm failing to  see the advantage of this approach creating 3 elements to make use of a single const containing possible operators

Could you explain this please?

File: api/src/modules/widgets/page-filters.service.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+export class PageFiltersService {
+  private filters: PageFilter[];
+
+  public constructor() {
+    this.filters = JSON.parse(
+      fs.readFileSync('data/filters/filters.json', 'utf-8'),
+    );
+  }
+
+  public async search(): Promise<PageFilter[]> {
```
Comment (alexeh) [inline api/src/modules/widgets/page-filters.service.ts:15]:
```suggestion
  public async getFilters(): Promise<PageFilter[]> {
```
Maybe would be a more comprehensive naming as we are not performing a lookup over the filters.
Then we could have another method to update the filters dynamically based on selections made in the FE

**Review by alexeh — COMMENTED** [review]
Looks cool @alepefe-vizz thanks! I have some doubts I would like to clarify before pushing this forward: (also left some comments)

Also, maybe `Filters` is more appropriate than `PageFilters`? Feels like it introduces a new concept `Page`. 

Conceptually, are filters applied to Sections, or are they applied to Widgets? We are updating the data within the widgets, but filters are actually applied globally to all sections (or more granularly), I am confused at this point. The naming and code organisation should go in this way I think, let me know what do you think and let's make an agreement. I would like @atrincas 's input on this point as well

File: shared/schemas/search-widget-data-params.schema.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { z } from 'zod';
+import {
+  VALID_SEARCH_WIDGET_DATA_OPERATORS,
+  WidgetDataFilterOperatorType,
+} from '@shared/dto/global/search-widget-data-params';
+
+export const SearchWidgetDataParamsSchema = z.object({
```
Comment (alexeh) [inline shared/schemas/search-widget-data-params.schema.ts:7]:
```suggestion
export const WidgetDataFilterSchema = z.object({
```

or 

```suggestion
export const SectionFilterSchema = z.object({
```

Feels like a better naming. What do you think?

File: api/data/filters/filters.json (hunk 1 above)
Comment (atrincas) [reply api/data/filters/filters.json:36]: @alexeh @alepefe-vizz I have already checked this with design, and the selected label is always in singular form.

File: shared/schemas/search-widget-data-params.schema.ts (hunk 5 above)
Comment (atrincas) [reply shared/schemas/search-widget-data-params.schema.ts:7]: I would go with WidgetDataFilterSchema

**Comment by atrincas** [comment]
My humble opinion as a more frontend dev, since we are using the `/sections` endpoint for fetching and filtering the data, I would place everything under the `sections Module`. That would be the first place I would look in the code. And I agree to just use the name `Filter`.

File: shared/schemas/search-widget-data-params.schema.ts (hunk 5 above)
Comment (alexeh) [reply shared/schemas/search-widget-data-params.schema.ts:7]: Sounds ok to me, but then we are conceptually tying the Filters to the Widgets, and not to the sections, take that into account

File: api/src/modules/widgets/page-filters.service.ts (hunk 4 above)
Comment (alepefe-vizz) [reply api/src/modules/widgets/page-filters.service.ts:15]: I think this is a matter of choice, I can change this if you prefer getFilters.

File: shared/dto/global/search-widget-data-params.ts (hunk 3 above)
Comment (alepefe-vizz) [reply shared/dto/global/search-widget-data-params.ts:11]: @alexeh maybe you need to implement something in the code and use `SEARCH_WIDGET_DATA_OPERATORS` as an enum. The second one is for a validation and the third the type?

What would you do?

File: shared/schemas/search-widget-data-params.schema.ts (hunk 2 above)
Comment (alepefe-vizz) [reply shared/schemas/search-widget-data-params.schema.ts:15]: @alexeh You mean like create the filter object and just having an array of the filter z.object?

Comment (alexeh) [reply shared/schemas/search-widget-data-params.schema.ts:15]: Not totally sure, but maybe use just a enum and avoid multiple level type casting. Just to simplify the code

File: shared/dto/global/search-widget-data-params.ts (hunk 3 above)
Comment (alexeh) [reply shared/dto/global/search-widget-data-params.ts:11]: can't we use just a enum? same as https://github.com/Vizzuality/4-growth/pull/115#discussion_r1780681440

File: api/src/modules/widgets/page-filters.service.ts (hunk 4 above)
Comment (alexeh) [reply api/src/modules/widgets/page-filters.service.ts:15]: It's just s suggestion, I don't have strong opinions. Go ahead if you feel seach is better :)

File: shared/dto/widgets/page-filter.dto.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+export type PageFilter = {
+  name: string;
+  isCustomFilter: boolean;
+  values: string[];
+  label: {
+    selected: string;
+    unSelected: string;
+  };
+};
```
Comment (atrincas) [inline shared/dto/widgets/page-filter.dto.ts:9]: proposal to change this to:
`type Filter = {
  name: string;
  values: string[];
  label: DefaultFilterLabels | string
};

type DefaultFilterLabels = {
  selected: string;
  unselected: string;
};`

**Review by alexeh — APPROVED** [review]
Nice job :)

File: shared/dto/widgets/page-filter.dto.ts (hunk 6 above)
Comment (alexeh) [reply shared/dto/widgets/page-filter.dto.ts]: Let's go into more details as we are able to pull actual data from somewhere

**Pull request merged**